tests/utils.py

```diff
@@ -270,3 +271,14 @@ def all_count(db):
 
 def row_count(db, cls):
     return db.session.query(cls).count()
+
+
+def redis_unavailable(cached_value=[]):
+    if len(cached_value) == 0:
+        try:
+            host = os.getenv('REDIS_HOST') or 'localhost'
+            redis.Redis(host=host).get('test')
+            cached_value.append(False)
+        except redis.exceptions.ConnectionError:
+            cached_value.append(True)
+    return cached_value[0]
```

Comment:
If cached value is a list, wouldn't it make more sense to return the last, most recent value rather than the first?

Reply:
`cached_value` has to be a list in order to retain the value.  There is only ever one item in the list so returning `cachd_value[0]` is ok.  To explain why `cached_value` has to be a list:

```python
def cached_value_not_list(cached_value=None):
    if cached_value is None:                      
        print('cached_value_not_list: not cached')
        cached_value = True
    return cached_value

                                          
def cached_value_is_list(cached_value=[]):
    if len(cached_value) == 0:                   
        print('cached_value_is_list: not cached')
        cached_value.append(True)
    return cached_value[0]

                       
cached_value_not_list()
# cached_value_not_list: not cached
cached_value_not_list()
# cached_value_not_list: not cached
cached_value_is_list()
# cached_value_is_list: not cached
cached_value_is_list()
```

In `cached_value_not_list`, we can't actually cache the value.  It is calculated every time.  The reason is because lists and dicts as keyword arguments are only initialized once, so if you modify the list in the function, it's actually changed.  (<- easy to have bugs if you're not aware of it)

Reply:
I'm really pleased that I asked the question as that answer was REALLY informative. 
This is exactly analogous to a static variable in C although less intuitive and .... oh yes I can see how that could lead to some really fun bugs.